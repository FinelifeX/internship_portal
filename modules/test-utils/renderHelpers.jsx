/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import TR from 'react-test-renderer';
import { render as RTLRender } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

/**
 * @typedef {Object.<string, any>} Props
 */

/**
 * @typedef {React.ReactNode} Children
 */

/**
 * @typedef {{props?: Props, children?: Children}} RenderOptions
 */

/**
 *
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @returns {JSX.Element}
 */
export const render = (Component, options = {}) => {
  const { props, children } = options;
  return <Component {...props}>{children}</Component>;
};

/**
 * Render React component inside MemoryRouter
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @param {import('react-router-dom').MemoryRouterProps} routerOptions
 * @returns {JSX.Element}
 */
export const renderWithRouter = (
  Component,
  options = {},
  routerOptions = {},
) => {
  const { props, children } = options;

  return (
    <MemoryRouter {...routerOptions}>
      <Component {...props}>{children}</Component>
    </MemoryRouter>
  );
};

/**
 * Shallow render component using "react-test-renderer"
 * @type {import('react-test-renderer').create}
 */
export const shallow = TR.create;

/**
 * Render component using "render()" from "@testing-library/react"
 * @type {import('@testing-library/react').render}
 */
export const mount = RTLRender;
