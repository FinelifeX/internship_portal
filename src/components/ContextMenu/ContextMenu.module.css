.menu {
  position: absolute;
  top: calc(100% + 8px);
  left: -6%;
  background-color: #fff;
  box-shadow: 0 0 12px rgba(0, 0, 0, 0.15);
  border-radius: 2px;
  z-index: 999;
}

.menuItems {
  list-style: none;
  padding: 0;
  margin: 0;
}

.menuButton {
  font-size: 14px;
  line-height: 20px;
  text-align: start;
  color: #595959;
}

.menuButton:hover,
.menuButton:focus,
.menuButton:active {
  background-color: rgba(1, 27, 74, 0.05);
  color: #1890ff;
}
