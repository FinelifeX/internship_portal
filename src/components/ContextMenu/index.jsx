import React, { useEffect } from 'react';
import PT from 'prop-types';
import classNames from 'classnames';
import Button from '../Button';
import styles from './ContextMenu.module.css';

const ContextMenu = ({ options, className, visible, onClose }) => {
  useEffect(() => {
    if (visible) {
      document.addEventListener('click', onClose);
    }

    return () => {
      document.removeEventListener('click', onClose);
    };
  }, [onClose, visible]);

  if (!visible) return null;

  return (
    <div className={classNames(styles.menu, className)}>
      <ul className={styles.menuItems}>
        {options.map(({ key, title, action, className: optionClassName }) => (
          <li key={key} className={styles.menuItem}>
            <Button
              className={classNames(styles.menuButton, optionClassName)}
              variant="link"
              onClick={action}
            >
              {title}
            </Button>
          </li>
        ))}
      </ul>
    </div>
  );
};

ContextMenu.defaultProps = {
  className: '',
};

ContextMenu.propTypes = {
  options: PT.arrayOf(
    PT.shape({
      key: PT.string,
      title: PT.node,
      action: PT.func,
      className: PT.string,
    }),
  ).isRequired,
  className: PT.string,
  visible: PT.bool.isRequired,
  onClose: PT.func.isRequired,
};

export default ContextMenu;
