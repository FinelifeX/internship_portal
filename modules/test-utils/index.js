import * as RenderHelpers from './renderHelpers';

export { RenderHelpers };
