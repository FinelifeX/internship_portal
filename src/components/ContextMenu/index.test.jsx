import { shallow, mount, render } from 'test-utils/renderHelpers';
import { fireEvent } from '@testing-library/react';
import ContextMenu from './index';

describe('ContextMenu component', () => {
  const TEST_ID_OPTION = 'option';
  const TEST_ID_OPTION_BUTTON = 'optionButton';
  const TEST_ID_CONTAINER = 'container';

  const createOption = (index) => ({
    key: `key${index}`,
    title: `Option ${index}`,
    action: jest.fn(),
  });
  const options = new Array(4).fill(0).map((item, idx) => createOption(idx));

  const defaultProps = {
    options,
    visible: true,
    onClose: jest.fn(),
  };

  const renderComponent = (props) =>
    render(ContextMenu, { props: { ...defaultProps, ...props } });

  it('should render correctly', () => {
    const tree = shallow(renderComponent()).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should not be rendered if visible=false', () => {
    const { queryByTestId } = mount(renderComponent({ visible: false }));

    expect(queryByTestId(TEST_ID_CONTAINER)).toBeNull();
  });

  it('should render all option elements', () => {
    const { getAllByTestId } = mount(renderComponent());

    expect(getAllByTestId(TEST_ID_OPTION)).toHaveLength(
      defaultProps.options.length,
    );
  });

  test('each option should call it\'s "action" callback when clicked', () => {
    const { getAllByTestId } = mount(renderComponent());

    getAllByTestId(TEST_ID_OPTION_BUTTON).forEach((item, idx) => {
      fireEvent.click(item);

      expect(defaultProps.options[idx].action).toBeCalledTimes(1);
    });
  });
});
