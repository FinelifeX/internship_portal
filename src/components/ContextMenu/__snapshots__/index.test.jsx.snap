// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu component should render correctly 1`] = `
<div
  className="menu"
  data-testid="container"
>
  <ul
    className="menuItems"
  >
    <li
      className="menuItem"
      data-testid="option"
    >
      <button
        className="button link menuButton"
        data-testid="button"
        onClick={[MockFunction]}
        type="button"
      >
        Option 0
      </button>
    </li>
    <li
      className="menuItem"
      data-testid="option"
    >
      <button
        className="button link menuButton"
        data-testid="button"
        onClick={[MockFunction]}
        type="button"
      >
        Option 1
      </button>
    </li>
    <li
      className="menuItem"
      data-testid="option"
    >
      <button
        className="button link menuButton"
        data-testid="button"
        onClick={[MockFunction]}
        type="button"
      >
        Option 2
      </button>
    </li>
    <li
      className="menuItem"
      data-testid="option"
    >
      <button
        className="button link menuButton"
        data-testid="button"
        onClick={[MockFunction]}
        type="button"
      >
        Option 3
      </button>
    </li>
  </ul>
</div>
`;
